### ISSUE TITLE
Change parameter names with set_section

### ISSUE THREAD

**Issue by itsdfish** [TuringLang/MCMCChains.jl#116]
Hi-

I am trying to modify the parameter names of a chain and run into the following error. Is there a way to do this? Thanks!

```
using MCMCChains
val = rand(500,5, 2)
chn = Chains(val, ["a", "b", "c", "d", "e"])
chn2 = set_section(chn, Dict(:parameters => ["A", "B", "C", "D", "E"]))
chn2
```

Error:

```
┌ Warning: Section mapping does not contain all parameter names, ["a", "b", "c", "d", "e"] assigned to :parameters.
└ @ MCMCChains ~/.julia/packages/MCMCChains/x3N5v/src/chains.jl:635
 chn2
  Object of type Chains, with data of type 500×5×2 Array{Float64,3}

Iterations        = 1:500
Thinning interval = 1
Chains            = 1, 2
Samples per chain = 500
parameters        = A, B, C, D, E, a, b, c, d, e

Error showing value of type Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}:
ERROR: ArgumentError: index ["A"] not found
Stacktrace:
 [1] axisindexes(::Type{AxisArrays.Categorical}, ::Array{String,1}, ::Array{String,1}) at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/indexing.jl:302
 [2] axisindexes at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/indexing.jl:174 [inlined]
 [3] macro expansion at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/indexing.jl:354 [inlined]
 [4] to_index at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/indexing.jl:311 [inlined]
 [5] getindex(::AxisArrays.AxisArray{Float64,3,Array{Float64,3},Tuple{AxisArrays.Axis{:iter,StepRange{Int64,Int64}},AxisArrays.Axis{:var,Array{String,1}},AxisArrays.Axis{:chain,Array{Int64,1}}}}, ::UnitRange{Int64}, ::Array{String,1}, ::Array{Int64,1}) at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/indexing.jl:117
 [6] getindex(::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}, ::UnitRange{Int64}, ::String, ::Int64) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/chains.jl:232
 [7] #ess#211(::Bool, ::Array{Symbol,1}, ::Int64, ::Missing, ::Function, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:271
 [8] #ess at ./none:0 [inlined]
 [9] #summarystats#226(::Bool, ::Bool, ::Array{Symbol,1}, ::Symbol, ::Missing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:403
 [10] (::getfield(StatsBase, Symbol("#kw##summarystats")))(::NamedTuple{(:showall, :sections, :etype, :digits),Tuple{Bool,Array{Symbol,1},Symbol,Missing}}, ::typeof(summarystats), ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at ./none:0
 [11] #describe#203(::Array{Float64,1}, ::Symbol, ::Bool, ::Array{Symbol,1}, ::Missing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Base.TTY, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:164
 [12] describe(::Base.TTY, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:164
 [13] #describe#202(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:136
 [14] describe at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/stats.jl:136 [inlined]
 [15] show(::IOContext{REPL.Terminals.TTYTerminal}, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/chains.jl:358
 [16] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at ./sysimg.jl:194
 [17] display(::REPL.REPLDisplay{REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Chains{Float64,Missing,NamedTuple{(:parameters,),Tuple{Array{String,1}}},NamedTuple{(:hashedsummary,),Tuple{Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}) at /Users/christopher.fisher/.julia/packages/OhMyREPL/GFHgr/src/output_prompt_overwrite.jl:6
 [18] display(::REPL.REPLDisplay, ::Any) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:135
 [19] display(::Any) at ./multimedia.jl:287
 [20] #invokelatest#1 at ./essentials.jl:742 [inlined]
 [21] invokelatest at ./essentials.jl:741 [inlined]
 [22] print_response(::IO, ::Any, ::Any, ::Bool, ::Bool, ::Any) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:155
 [23] print_response(::REPL.AbstractREPL, ::Any, ::Any, ::Bool, ::Bool) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:140
 [24] (::getfield(REPL, Symbol("#do_respond#38")){Bool,getfield(Atom, Symbol("##172#173")),REPL.LineEditREPL,REPL.LineEdit.Prompt})(::Any, ::Any, ::Any) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:714
 [25] #invokelatest#1 at ./essentials.jl:742 [inlined]
 [26] invokelatest at ./essentials.jl:741 [inlined]
 [27] run_interface(::REPL.Terminals.TextTerminal, ::REPL.LineEdit.ModalInterface, ::REPL.LineEdit.MIState) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/LineEdit.jl:2273
 [28] run_frontend(::REPL.LineEditREPL, ::REPL.REPLBackendRef) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:1035
 [29] run_repl(::REPL.AbstractREPL, ::Any) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:192
 [30] (::getfield(Base, Symbol("##734#736")){Bool,Bool,Bool,Bool})(::Module) at ./client.jl:362
 [31] #invokelatest#1 at ./essentials.jl:742 [inlined]
 [32] invokelatest at ./essentials.jl:741 [inlined]
 [33] run_main_repl(::Bool, ::Bool, ::Bool, ::Bool, ::Bool) at ./client.jl:346
 [34] exec_options(::Base.JLOptions) at ./client.jl:284
 [35] _start() at ./client.jl:436
```

**Comment by cpfiffer** [comment]
There is not currently a way to do this in a specific function, no. You might be able to construct a new `Chains` object with a different name and map, try

```julia
chn2 = Chains(chn.value.data, ["A", "B", "C", "D", "E"])
```

If your chain is made with Turing, sometimes the `Chain.info` field contains some important stuff. You might also really want the `logevidence` field. You can transfer these over to the new chain with the keyword arguments

```julia
chn2 = Chains(chn.value.data, ["A", "B", "C", "D", "E"], info=chn.info, logevidence=chn.logevidence)
```

**Comment by itsdfish** [comment]
Thanks for your reply. One of the problems I am having is that chain.value.data include internal and parameter samples. This makes it difficult to change names when both are present. Here is my attempt to flexibly change names: 



```
function changeNames(chain,newNames)
    parms = chain.name_map.parameters
    parms = replace(parms,newNames...)
    internals = chain.name_map.internals
    internals = replace(internals,newNames...)
    return Chains(chain.value.data,parms,info=chain.info,Dict(:internals=>internals))
end

# Load Turing and MCMCChains.
using Turing, MCMCChains,Distributions


@model coinflip(y) = begin
    
    # Our prior belief about the probability of heads in a coin.
    p ~ Beta(1, 1)
    
    # The number of observations.
    N = length(y)
    for n in 1:N
        # Heads or tails of a coin are drawn from a Bernoulli distribution.
        y[n] ~ Bernoulli(p)
    end
end

data = rand(Bernoulli(.5),10)

# Settings of the Hamiltonian Monte Carlo (HMC) sampler.
iterations = 1000
ϵ = 0.05
τ = 10

# Start sampling.
chain = sample(coinflip(data), HMC(iterations, ϵ, τ));
newNames = Dict("p"=>"θ")
newChain = changeNames(chain,newNames)
```

Error:

```
ERROR: ArgumentError: the length of each axis must match the corresponding size of data
Stacktrace:
 [1] AxisArrays.AxisArray(::Array{Union{Missing, Float64},3}, ::Tuple{AxisArrays.Axis{:iter,StepRange{Int64,Int64}},AxisArrays.Axis{:var,Array{String,1}},AxisArrays.Axis{:chain,Array{Int64,1}}}) at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/core.jl:222
 [2] AxisArrays.AxisArray(::Array{Union{Missing, Float64},3}, ::AxisArrays.Axis{:iter,StepRange{Int64,Int64}}, ::Vararg{Union{Axis, AbstractArray{T,1} where T},N} where N) at /Users/christopher.fisher/.julia/packages/AxisArrays/G6pZY/src/core.jl:215
 [3] #Chains#24(::Int64, ::Int64, ::Missing, ::NamedTuple{(:samples, :hashedsummary),Tuple{Array{Turing.Utilities.Sample,1},Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}, ::Type, ::Array{Union{Missing, Float64},3}, ::Array{String,1}, ::Dict{Symbol,Array{String,1}}) at /Users/christopher.fisher/.julia/packages/MCMCChains/x3N5v/src/chains.jl:134
 [4] (::getfield(Core, Symbol("#kw#Type")))(::NamedTuple{(:info,),Tuple{NamedTuple{(:samples, :hashedsummary),Tuple{Array{Turing.Utilities.Sample,1},Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}}, ::Type{Chains}, ::Array{Union{Missing, Float64},3}, ::Array{String,1}, ::Dict{Symbol,Array{String,1}}) at ./none:0
 [5] changeNames(::Chains{Union{Missing, Float64},Float64,NamedTuple{(:internals, :parameters),Tuple{Array{String,1},Array{String,1}}},NamedTuple{(:samples, :hashedsummary),Tuple{Array{Turing.Utilities.Sample,1},Base.RefValue{Tuple{UInt64,Array{ChainDataFrame,1}}}}}}, ::Dict{String,String}) at ./untitled-e1651d272195110f9c4fbdb6cac219fa:6
 [6] top-level scope at none:0
```

**Comment by cpfiffer** [comment]
Probably the best I can do is give you a utility function that accepts a `Dict` of `old -> new` names to replace. Changing names by section I think is kind of a strange functionality to support.

Can I ask why you're changing the names? Why not just use θ in the first place?

**Comment by goedman** [comment]
Hi Cameron,

One issue is that Stan does not accept `greek` names and in MCMCBenchmarks we compare Turing, DynamicHMC and Stan NUTS variant results.

Best,
Rob

**Comment by cpfiffer** [comment]
Fixed in #119. Let me know if you guys need better/different functionality.
